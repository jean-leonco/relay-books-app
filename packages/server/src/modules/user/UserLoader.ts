import { mongooseLoader } from '@entria/graphql-mongoose-loader';
import DataLoader from 'dataloader';

import { Types } from 'mongoose';

import { DataLoaderKey, GraphQLContext, LoggedGraphQLContext } from '../../types';

import { isLoggedIn } from '../../core/security';

import UserModel, { IEmailSchema, IUser } from './UserModel';

export default class User {
  public registeredType = 'User';

  id: string;
  _id: Types.ObjectId;
  name: string;
  surname: string;
  email: IEmailSchema;
  lang?: string;
  isActive: boolean;
  removedAt: Date | null;
  createdAt: Date;
  updatedAt: Date;

  constructor(data: IUser, context: GraphQLContext) {
    this.id = data.id || data._id;
    this._id = data._id;
    this.removedAt = data.removedAt;
    this.createdAt = data.createdAt;
    this.updatedAt = data.updatedAt;
    this.isActive = data.isActive;
    this.name = data.name;
    this.surname = data.surname;
    this.email = data.email;
    this.lang = data.lang;
  }
}

const viewerCanSee = (context: GraphQLContext, data: IUser) => {
  if (!isLoggedIn(context)) {
    return false;
  }

  return true;
};

export const getLoader = () => new DataLoader<DataLoaderKey, IUser>((ids) => mongooseLoader(UserModel, ids));

export const load = async (context: GraphQLContext, id: DataLoaderKey, isLogin?: boolean) => {
  if (!id) {
    return null;
  }

  try {
    const data = await context.dataloaders.UserLoader.load(id);

    if (!data) {
      return null;
    }

    if (isLogin) {
      return new User(data, context);
    }

    return viewerCanSee(context, data) ? new User(data, context) : null;
  } catch (err) {
    return null;
  }
};

export const clearCache = ({ dataloaders }: GraphQLContext, id: Types.ObjectId) =>
  dataloaders.UserLoader.clear(id.toString());

export const primeCache = ({ dataloaders }: GraphQLContext, id: Types.ObjectId, data: IUser) =>
  dataloaders.UserLoader.prime(id.toString(), data);

export const clearAndPrimeCache = (context: GraphQLContext, id: Types.ObjectId, data: IUser) =>
  clearCache(context, id) && primeCache(context, id, data);

export const findUserByEmail = async (context: GraphQLContext, email: string, raw = true): Promise<IUser | null> => {
  const result = raw
    ? await UserModel.findOne({ 'email.email': email }).lean<IUser>()
    : await UserModel.findOne({ 'email.email': email });

  return result;
};

export const userEmailExists = async (context: GraphQLContext, email: string): Promise<number | null> => {
  return await UserModel.countDocuments({ 'email.email': email });
};

export const meEmailExists = async (context: LoggedGraphQLContext, email: string): Promise<number | null> => {
  const { user } = context;
  return await UserModel.countDocuments({ 'email.email': email, _id: { $ne: user._id } });
};
