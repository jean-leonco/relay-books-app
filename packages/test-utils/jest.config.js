const pkg = require('./package');

module.exports = {
  rootDir: './',
  name: pkg.name,
  displayName: pkg.name.toUpperCase(),
  testPathIgnorePatterns: ['/node_modules/', './build'],
  coverageReporters: ['lcov', 'html'],
  setupFilesAfterEnv: ['<rootDir>/test/setupTestFramework.js'],
  globalSetup: '<rootDir>/test/setup.js',
  globalTeardown: '<rootDir>/test/teardown.js',
  transform: {
    // '^.+\\.(js|jsx|ts|tsx)?$': require.resolve('babel-jest'),
    '^.+\\.(js|jsx|ts|tsx)?$': '<rootDir>/test/babel-transformer',
  },
};
