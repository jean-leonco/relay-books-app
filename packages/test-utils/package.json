{
  "name": "@booksapp/test-utils",
  "version": "0.0.1",
  "private": true,
  "license": "UNLICENSED",
  "main": "src/index.ts",
  "module": "src/index.ts",
  "scripts": {
    "test": "jest",
    "test:watch": "jest --watch"
  },
  "dependencies": {
    "@booksapp/babel": "^0.0.1",
    "@booksapp/types": "^0.0.1",
    "@koa/router": "^9.3.1",
    "babel-jest": "^26.1.0",
    "graphql-relay": "0.6.0",
    "koa": "^2.13.0",
    "koa-bodyparser": "^4.3.0",
    "mongoose": "^5.10.7",
    "supertest": "^5.0.0"
  }
}
