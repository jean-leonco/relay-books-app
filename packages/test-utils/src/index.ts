export { defaultFrozenKeys, sanitizeValue, sanitizeTestObject } from './objectNormalization';
export { httpRequestGraphql } from './httpRequestGraphql';
