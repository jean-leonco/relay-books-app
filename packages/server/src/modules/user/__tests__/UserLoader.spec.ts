import { sanitizeTestObject } from '@booksapp/test-utils';

import {
  clearDbAndRestartCounters,
  connectMongoose,
  createUser,
  disconnectMongoose,
  getContext,
} from '../../../../test/helpers';

import * as UserLoader from '../UserLoader';

beforeAll(connectMongoose);

beforeEach(clearDbAndRestartCounters);

afterAll(disconnectMongoose);

describe('UserLoader functions', () => {
  it('should verify that e-mail already exists', async () => {
    const email = 'jean@gmail.io';
    const user = await createUser({ email: { email, wasVerified: true } });
    const context = await getContext({ user });

    const result = await UserLoader.userEmailExists(context, email);

    expect(result).toBe(1);
  });

  it('should verify that e-mail does not exists', async () => {
    const email = 'jean@gmail.io';
    const user = await createUser({ email: { email, wasVerified: true } });
    const context = await getContext({ user });

    const result = await UserLoader.userEmailExists(context, email + '.br');

    expect(result).toBe(0);
  });

  it('should find user by e-mail', async () => {
    const email = 'jean@gmail.io';
    const user = await createUser({ email: { email, wasVerified: true } });
    const context = await getContext({ user });

    const result = await UserLoader.findUserByEmail(context, email);

    expect(sanitizeTestObject(result)).toMatchSnapshot();
  });
});
