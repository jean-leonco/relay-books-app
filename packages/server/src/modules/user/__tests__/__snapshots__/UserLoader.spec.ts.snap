// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserLoader functions should find user by e-mail 1`] = `
Object {
  "__v": 0,
  "_id": "ObjectId",
  "createdAt": "FROZEN-CREATEDAT",
  "email": Object {
    "email": "jean@gmail.io",
    "wasVerified": true,
  },
  "isActive": true,
  "name": "User #1",
  "password": "FROZEN-PASSWORD",
  "removedAt": "EMPTY",
  "updatedAt": "FROZEN-UPDATEDAT",
}
`;
